### ISSUE TITLE
🔴 The icons for MKV and WEBM format videos are blank.

### ISSUE THREAD

**Issue by Thomas-Harrell** [iina/iina#4491]
**System and IINA version:**

- macOS 11.7.2
- IINA 1.3.2

**Expected behavior:**

- The icons are displayed normally.

**Actual behavior:**

- The icons are displayed abnormally.

![01](https://github.com/iina/iina/assets/106761803/c8861e17-3574-42c7-9044-ddc512ca31c9)

<details>
<summary><b>Crash report:</b></summary>

```
```
</details>

<br>

<details>
<summary><b>mpv log:</b></summary>

```
```
</details>

<br>

**Steps to reproduce:**

- Use `qlmanage -r cache && qlmanage -r` to clear the icon cache in Finder.

- After that, you will see the icons for MKV and WEBM format videos are blank.

- [x] MPV does not have this problem.

**How often does this happen?**

- Since the update of version 1.3.2 of IINA

**Comment by uiryuu** [comment]
I was unable to reproduce the issue. I've tried to run `qlmanage -r cache && qlmanage -r` and restarted Finder, I can still see the icon for webm. Do I need to restart or something?

**Comment by Thomas-Harrell** [comment]
> I was unable to reproduce the issue. I've tried to run `qlmanage -r cache && qlmanage -r` and restarted Finder, I can still see the icon for webm. Do I need to restart or something?

- Thank you for your reply.

- My previous description should have been more detailed.

- It might be because I am still using a system from the Big Sur version.

- https://github.com/iina/iina/discussions/3670

- I encountered this issue in the nightly build versions before. I thought this issue would be resolved in the official version.

- https://nightly.iina.io

- I have also seen some similar feedback.

- https://github.com/iina/iina/issues/4030

- https://github.com/iina/iina/issues/4321

- `qlmanage -r cache && qlmanage -r && killall Finder`

- I used the command above just to quickly refresh the Finder's icon cache because I found that if the icon is not refreshed immediately, the icon will still exist for a period of time.

- However, after restarting the system, the icon cache may disappear. Or, after opening a file, you can see that the icon is blank in the top bar of the IINA window.

**Comment by Fujiwara-Chika** [comment]
I can confirm that this is a bug in macOS since Catalina, to fix it you can move the IINA.app to the trash, then reboot the system and drag it back.

**Comment by AlejandroVolkova** [comment]
this issue happened to me too, but as @Fujiwara-Chika said it's most likely a macOS bug, definitely not isolated to IINA, the way macOS manages the icon cache has changed over time and it sucks now.

However you can fix this issue by : 

- chaging the `default app` to another one 
- selecting `IINA (as default)` again
- then pressing the box `Change All...` 
- restart (maybe it takes a few restarts but it works.)

Example : 

<img width="377" alt="Screenshot 2023-07-15 at 9 16 57 PM" src="https://github.com/iina/iina/assets/120342552/8efacbcf-10d2-4335-b200-4fb1fea88c84">

**Comment by Thomas-Harrell** [comment]
> this issue happened to me too, but as @Fujiwara-Chika said it's most likely a macOS bug, definitely not isolated to IINA, the way macOS manages the icon cache has changed over time and it sucks now.
> 
> However you can fix this issue by :
> 
> * chaging the `default app` to another one
> * selecting `IINA (as default)` again
> * then pressing the box `Change All...`
> * restart (maybe it takes a few restarts but it works.)
> 
> Example :
> 
> <img alt="Screenshot 2023-07-15 at 9 16 57 PM" width="377" src="https://user-images.githubusercontent.com/120342552/253780733-8efacbcf-10d2-4335-b200-4fb1fea88c84.png">

- Thank you for your response.

- The command `qlmanage -r cache && qlmanage -r && killall Finder` can immediately refresh the icon cache.

- In version 1.3.1 of IINA, the icons can be refreshed, but in version 1.3.2, only blank icons are displayed.

- I am using the Big Sur operating system.

**Comment by Fujiwara-Chika** [comment]
> > this issue happened to me too, but as @Fujiwara-Chika said it's most likely a macOS bug, definitely not isolated to IINA, the way macOS manages the icon cache has changed over time and it sucks now.
> > However you can fix this issue by :
> > 
> > * chaging the `default app` to another one
> > * selecting `IINA (as default)` again
> > * then pressing the box `Change All...`
> > * restart (maybe it takes a few restarts but it works.)
> > 
> > Example :
> > <img alt="Screenshot 2023-07-15 at 9 16 57 PM" width="377" src="https://user-images.githubusercontent.com/120342552/253780733-8efacbcf-10d2-4335-b200-4fb1fea88c84.png">
> 
> * Thank you for your response.
> * The command `qlmanage -r cache && qlmanage -r && killall Finder` can immediately refresh the icon cache.
> * In version 1.3.1 of IINA, the icons can be refreshed, but in version 1.3.2, only blank icons are displayed.
> * I am using the Big Sur operating system.

Actually there are two versions of IINA 1.3.2, in the initial version there was indeed a bug with blank icons, this was quickly fixed in subsequent version, but if you happen to have installed the first version then you need to regenerate the cache again as I said above, that command doesn't work.

**Comment by Thomas-Harrell** [comment]
> > > this issue happened to me too, but as @Fujiwara-Chika said it's most likely a macOS bug, definitely not isolated to IINA, the way macOS manages the icon cache has changed over time and it sucks now.
> > > However you can fix this issue by :
> > > 
> > > * chaging the `default app` to another one
> > > * selecting `IINA (as default)` again
> > > * then pressing the box `Change All...`
> > > * restart (maybe it takes a few restarts but it works.)
> > > 
> > > Example :
> > > <img alt="Screenshot 2023-07-15 at 9 16 57 PM" width="377" src="https://user-images.githubusercontent.com/120342552/253780733-8efacbcf-10d2-4335-b200-4fb1fea88c84.png">
> > 
> > 
> > 
> > * Thank you for your response.
> > * The command `qlmanage -r cache && qlmanage -r && killall Finder` can immediately refresh the icon cache.
> > * In version 1.3.1 of IINA, the icons can be refreshed, but in version 1.3.2, only blank icons are displayed.
> > * I am using the Big Sur operating system.
> 
> Actually there are two versions of IINA 1.3.2, in the initial version there was indeed a bug with blank icons, this was quickly fixed in subsequent version, but if you happen to have installed the first version then you need to regenerate the cache again as I said above, that command doesn't work.

- Now I know that there are actually two versions of IINA 1.3.2.

- I updated to the latest version at the time of the 1.3.2 release.

- I didn't expect to encounter such a problem.

- After I installed the second version of IINA 1.3.2 and used the `qlmanage -r cache && qlmanage -r && killall Finder` command, I can now see the icons for MKV and WEBM formats. They are no longer displaying as blank icons.

- Thank you for your reminder.

![01](https://github.com/iina/iina/assets/106761803/978a65d5-0398-4e34-b67a-e422bc507ce6)

![02](https://github.com/iina/iina/assets/106761803/27923afb-beea-4bf0-9ff8-cc58328ed3d4)

![01-1](https://github.com/iina/iina/assets/106761803/f8d9a858-f8ac-4a36-bf41-a205c29f82e2)

![02-1](https://github.com/iina/iina/assets/106761803/b61aa29e-2305-4bd8-8c16-e9c24d56e8e5)

![03](https://github.com/iina/iina/assets/106761803/bc98447e-05f4-41e2-ac22-a67544a74bae)

![04](https://github.com/iina/iina/assets/106761803/36c70720-6754-48e6-903f-813f39606801)

**Comment by Mikasa-san** [comment]
Guys I have the same issue and I have MacOS Sonoma beta 3

Edit: I used `qlmanage -r cache && qlmanage -r && killall Finder` and it fixed now.

**Comment by uiryuu** [comment]
Thanks for confirming the issue. Closed as no issue exists in the latest release version

**Comment by dog-sled-archive** [comment]
I did some poking into this and found the cause-- `doc_mkv.icns` is referenced in `Info.plist` but never shipped in the app bundle.

On 1.4.3 Build 167:

```
/usr/libexec/PlistBuddy -c "Print CFBundleDocumentTypes" /Applications/IINA.app/Contents/Info.plist
```

Matroska entry includes `CFBundleTypeIconFile = doc_mkv`, which tells Launch Services to look for `doc_mkv.icns`. 

```
ls /Applications/IINA.app/Contents/Resources/*.icns
```

Only `AppIcon.icns` is present. So `doc_mkv.icns` doesn't exist. Launch Services finds the declaration, finds no file, and falls back to generic icon.

Maybe related: the `CFBundleDocumentTypes` entry for Matroska uses `CFBundleTypeExtensions` rather than `LSItemContentTypes` pointing to the declared `io.iina.mkv` UTI. Declarations are disconnected.

I ran `qlmanage -r cache && qlmanage -r && killall Finder` to rule out a cache issue. No change in my case (1.4.3 on Tahoe).

Shipping `doc_mkv.icns` in `Resources/` should fix. Happy to help test if a build is available.
